libraries/AP_ExternalAHRS/AP_ExternalAHRS_AdvancedNavigation.cpp

```diff
@@ -0,0 +1,1869 @@
+        CERTUS                = 26,
+        ARIES,
+        BOREAS_D90,
+        BOREAS_D90_FPGA       = 35,
+        BOREAS_COIL,
+        CERTUS_MINI_A         = 49,
+        CERTUS_MINI_N,
+        CERTUS_MINI_D,
+    };
+
+    uint32_t software_version;
+    AN_DeviceId device_id;
+    uint32_t hardware_revision;
+    uint32_t serial_1;
+    uint32_t serial_2;
+    uint32_t serial_3;
+};
+
+/*
+    ANPP System State Packet (20)
+    
+    Provides the EKF output of the vehicle state
+    
+    https://docs.advancednavigation.com/certus/ANPP/SystemStatePacket.htm
+*/
+struct PACKED AN_SystemState
+{
+    uint16_t system_status;
+    uint16_t filter_status;
+    uint32_t unix_time_seconds;
+    uint32_t microseconds;
+    double llh[3];                   // rad,rad,m
+    float velocity_ned[3];           // m/s
+    float body_acceleration[3];      // m/s/s
+    float g_force;                   // g's
+    float rph[3];                    // rad
+    float angular_velocity[3];       // rad/s
+    float llh_standard_deviation[3]; // m
+
+    enum AN_SystemStatusFlags
+    {
+        AN_SYSTEM_FAILURE               = 1 << 0,
+        AN_ACCELEROMETER_SENSOR_FAILURE = 1 << 1,
+        AN_GYROSCOPE_SENSOR_FAILURE     = 1 << 2,
+        AN_MAGNETOMETER_SENSOR_FAILURE  = 1 << 3,
+        AN_PRESSURE_SENSOR_FAILURE      = 1 << 4,
+        AN_GNSS_FAILURE                 = 1 << 5,
+        AN_ACCELEROMETER_OVER_RANGE     = 1 << 6,
+        AN_GYROSCOPE_OVER_RANGE         = 1 << 7,
+        AN_MAGNETOMETER_OVER_RANGE      = 1 << 8,
+        AN_PRESSURE_OVER_RANGE          = 1 << 9,
+        AN_MINIMUM_TEMPERATURE_ALARM    = 1 << 10,
+        AN_MAXIMUM_TEMPERATURE_ALARM    = 1 << 11,
+        AN_INTERNAL_DATA_LOGGING_ERROR  = 1 << 12,
+        AN_HIGH_VOLTAGE_ALARM           = 1 << 13,
+        AN_GNSS_ANTENNA_FAULT           = 1 << 14,
+        AN_SERIAL_PORT_OVERFLOW_ALARM   = 1 << 15
+    };
+
+    enum AN_FilterStatusFlags
+    {
+        AN_ORIENTATION_FILTER_INITIALISED = 1 << 0,
+        AN_INS_FILTER_INITIALISED         = 1 << 1,
+        AN_HEADING_INITIALISED            = 1 << 2,
+        AN_UTC_TIME_INITIALISED           = 1 << 3,
+
+                                          // Bits 4,5,6 = gnss_fix_type (3-bit field)
+        AN_GNSS_FIX_TYPE_MASK  = 0x0070,  // bits 4–6
+        AN_GNSS_FIX_TYPE_SHIFT = 4,       // to isolate/shift value
+
+        AN_EVENT1_FLAG                  = 1 << 7,
+        AN_EVENT2_FLAG                  = 1 << 8,
+        AN_INTERNAL_GNSS_ENABLED        = 1 << 9,
+        AN_DUAL_ANTENNA_HEADING_ACTIVE  = 1 << 10,
+        AN_VELOCITY_HEADING_ENABLED     = 1 << 11,
+        AN_ATMOSPHERIC_ALTITUDE_ENABLED = 1 << 12,
+        AN_EXTERNAL_POSITION_ACTIVE     = 1 << 13,
+        AN_EXTERNAL_VELOCITY_ACTIVE     = 1 << 14,
+        AN_EXTERNAL_HEADING_ACTIVE      = 1 << 15
+    };
+    
+    void get_location(Location &loc) const
+    {
+        loc.lat =  (int32_t)(this->llh[0] * RAD_TO_DEG_DOUBLE * 1.0e7);
```

Comment:
this should be done with a Location constructor, change_alt_frame() will try to map from current frame (eg. terrain lookup)